### PR TITLE
chore: Improves linting rule to catch mishandled promises

### PR DESCRIPTION
**Pull request by tommoor**
Unfortunately having this lint rule work properly required upgrading typescript and eslint.

### CODE REVIEW

File: server/routes/api/documents/documents.test.ts [hunk 1]
```diff
@@ -1573,21 +1573,18 @@ describe("#documents.search", () => {
...
-          },
-        });
-        expect(searchQuery.length).toBe(1);
-        expect(searchQuery[0].results).toBe(0);
-        expect(searchQuery[0].source).toBe("app");
-        resolve(undefined);
-      }, 100);
+    // setTimeout is needed here because SearchQuery is saved asynchronously
+    // in order to not slow down the response time.
+    await new Promise((resolve) => setTimeout(resolve, 100));
```
Comment (apoorv-mishra) [inline server/routes/api/documents/documents.test.ts:1578]: We already have a util for this in `server/utils/timers.ts`

File: app/scenes/Search/components/UserFilter.tsx [hunk 2]
```diff
@@ -4,6 +4,7 @@ import * as React from "react";
 import { useTranslation } from "react-i18next";
 import styled from "styled-components";
 import Avatar from "~/components/Avatar";
+import { AvatarSize } from "~/components/Avatar/Avatar";
```
Comment (apoorv-mishra) [inline app/scenes/Search/components/UserFilter.tsx:7]: These changes are unrelated to linting?

File: tsconfig.json [hunk 3]
```diff
@@ -21,6 +21,7 @@
     "strictNullChecks": true,
     "noEmit": true,
     "skipLibCheck": true,
+    "ignoreDeprecations": "5.0",
```
Comment (apoorv-mishra) [inline tsconfig.json:24]: Remember this as a temporary fix.

**Comment by apoorv-mishra** [comment]
Oops! This is already merged :laughing:

File: server/routes/api/documents/documents.test.ts (hunk 1 above)
Comment (tommoor) [reply server/routes/api/documents/documents.test.ts:1578]: gtk, I don't feel too bad about a dupe here with it being tests
